mergify_engine/merge_fast_forwarder.py

```diff
@@ -0,0 +1,120 @@
+# -*- encoding: utf-8 -*-
+#
+# Copyright © 2022 Mergify SAS
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import dataclasses
+import typing
+
+from mergify_engine import exceptions
+from mergify_engine import gitter
+
+
+if typing.TYPE_CHECKING:
+    from mergify_engine import context
+    from mergify_engine.dashboard import user_tokens
+
+
+@dataclasses.dataclass
+class MergeFastForwardUnexpectedError(Exception):
+    reason: str
+
+
+@dataclasses.dataclass
+class MergeFastForwardNeedRetry(exceptions.EngineNeedRetry):
+    reason: str
+
+
+@dataclasses.dataclass
+class MergeFastForwardFailed(Exception):
+    reason: str
+
+
+GIT_MESSAGE_TO_EXCEPTION = {
+    "(non-fast-forward)": MergeFastForwardFailed,
+    "Updates were rejected because the tip of your current branch is behind": MergeFastForwardFailed,
+    "You may want to first integrate the remote changes": MergeFastForwardFailed,
+    "couldn't find remote ref": MergeFastForwardFailed,
+}
+
+
+async def doit(
+    ctxt: "context.Context",
+    bot_account_user: typing.Optional["user_tokens.UserTokensUser"],
+) -> None:
+    if ctxt.pull["head"]["repo"] is None:
+        raise MergeFastForwardFailed("The head repository does not exists anymore")
+
+    head_branch = ctxt.pull["head"]["ref"]
+    base_branch = ctxt.pull["base"]["ref"]
+    git = gitter.Gitter(ctxt.log)
+    try:
+        await git.init()
+
+        if bot_account_user is None:
+            token = await ctxt.client.get_access_token()
+            await git.configure()
+            username = "x-access-token"
+            password = token
+        else:
+            raise RuntimeError("I don't think this works")
+            await git.configure(
+                bot_account_user["name"] or bot_account_user["login"],
+                bot_account_user["email"],
+            )
+            username = bot_account_user["oauth_access_token"]
+            password = ""  # nosec
+
+        await git.setup_remote("head", ctxt.pull["head"]["repo"], username, password)
+        await git.setup_remote("base", ctxt.pull["base"]["repo"], username, password)
+
+        await git("fetch", "--quiet", "head", head_branch)
+        await git("fetch", "--quiet", "base", base_branch)
```

Comment:
That sounds like it's going to download everything?
Do we have to fetch? promisor should be enough? 🤞

Reply:
you need to fetch everything except blobs/files, promisor do it for us.